### PR TITLE
Use a `Set<T>` instead of a `Map<T, bool>`

### PR DESCRIPTION
**Pull request by oli-obk**
r? @nikomatsakis 

introduced in #44501

### CODE REVIEW

File: src/librustc_passes/consts.rs [hunk 1]
```diff
@@ -79,12 +79,12 @@ fn const_is_rvalue_promotable_to_static<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                      .expect("rvalue_promotable_map invoked with non-local def-id");
     let body_id = tcx.hir.body_owned_by(node_id);
     let body_hir_id = tcx.hir.node_to_hir_id(body_id.node_id);
-    tcx.rvalue_promotable_map(def_id).contains_key(&body_hir_id.local_id)
+    tcx.rvalue_promotable_map(def_id).contains(&body_hir_id.local_id)
```
Comment (oli-obk) [inline src/librustc_passes/consts.rs]: I did this exact thing in clippy, and it was very wrong. I'm fairly certain it was wrong here, too.

This checked whether the `local_id` was ever checked for rvalue promotion. Not whether it is rvalue promotable.

File: src/librustc/middle/mem_categorization.rs [hunk 2]
```diff
@@ -897,7 +897,7 @@ impl<'a, 'gcx, 'tcx> MemCategorizationContext<'a, 'gcx, 'tcx> {
                            expr_ty: Ty<'tcx>)
                            -> cmt<'tcx> {
         let hir_id = self.tcx.hir.node_to_hir_id(id);
-        let promotable = self.rvalue_promotable_map.as_ref().map(|m| m[&hir_id.local_id])
+        let promotable = self.rvalue_promotable_map.as_ref().map(|m| m.contains(&hir_id.local_id))
```
Comment (oli-obk) [inline src/librustc/middle/mem_categorization.rs]: This is the only change in semantics. We do not panic anymore if `local_id` was never checked for rvalue promotion, we just return `false`

Comment (nikomatsakis) [reply src/librustc/middle/mem_categorization.rs]: yeah, seems..ok

File: src/librustc_passes/consts.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_passes/consts.rs]: 👍

**Comment by nikomatsakis** [comment]
@oli-obk I wonder if we can make a test for that `contains_key` call -- I agree it was wrong, interesting that it didn't break anything.

**Comment by oli-obk** [comment]
I wondered the same thing, but found it very hard to actually do. There are too many other checks, it was probably redundant anyway, since it would additionally have to go through trans const eval

**Pull request merged**